enterprise/dbcrypt/dbcrypt.go

```diff
@@ -0,0 +1,296 @@
+// Package dbcrypt provides a database.Store wrapper that encrypts/decrypts
+// values stored at rest in the database.
+//
+// Encryption is done using a Cipher. The Cipher is stored in an atomic pointer
+// so that it can be rotated as required.
+//
+// The Cipher is currently used to encrypt/decrypt the following fields:
+// - database.UserLink.OAuthAccessToken
+// - database.UserLink.OAuthRefreshToken
+// - database.GitAuthLink.OAuthAccessToken
+// - database.GitAuthLink.OAuthRefreshToken
+// - database.DBCryptSentinelValue
+//
+// Encrypted fields are stored in the following format:
+// "dbcrypt-${b64encode(<first 7 digits of cipher's SHA256 digest>-<encrypted value>)}"
+//
+// The first 7 characters of the cipher's SHA256 digest are used to identify the cipher
+// used to encrypt the value.
+//
+// Multiple ciphers can be provided to support key rotation. The primary cipher is used
+// to encrypt and decrypt all data. Secondary ciphers are only used for decryption.
+// We currently only use a single secondary cipher.
+package dbcrypt
+
+import (
+	"context"
+	"database/sql"
+	"encoding/base64"
+	"errors"
+	"strings"
+
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbauthz"
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+)
+
+// MagicPrefix is prepended to all encrypted values in the database.
+// This is used to determine if a value is encrypted or not.
+// If it is encrypted but a key is not provided, an error is returned.
+// MagicPrefix will be followed by the first 7 characters of the cipher's
+// SHA256 digest, followed by a dash, followed by the base64-encoded
+// encrypted value.
+const MagicPrefix = "dbcrypt-"
+
+// sentinelValue is the value that is stored in the database to indicate
+// whether encryption is enabled. If not enabled, the value either not
+// present, or is the raw string "coder".
+// Otherwise, the value must be the encrypted value of the string "coder"
+// using the current cipher.
+const sentinelValue = "coder"
+
+var (
+	ErrNotEnabled       = xerrors.New("encryption is not enabled")
+	ErrSentinelMismatch = xerrors.New("database is already encrypted under a different key")
+	b64encode           = base64.StdEncoding.EncodeToString
+	b64decode           = base64.StdEncoding.DecodeString
+)
+
+// DecryptFailedError is returned when decryption fails.
+// It unwraps to sql.ErrNoRows.
+type DecryptFailedError struct {
+	Inner error
+}
+
+func (e *DecryptFailedError) Error() string {
+	return xerrors.Errorf("decrypt failed: %w", e.Inner).Error()
+}
+
+func (*DecryptFailedError) Unwrap() error {
+	return sql.ErrNoRows
```

Comment:
What's the logic behind making failed decryption look like `sql.ErrNoRows`?

I feel like it will mask operational issues (like having the wrong key), making it take longer for ops to find and fix the problem.

Reply:
I don't have a good answer for that apart from "it felt like the right thing to do"?
But you're correct, making it unwrap to ErrNoRows does mask the case where we can't decrypt a value. In the case of an oauth token, it just means that the user has to re-authenticate. But for other data, we could end up in a weird situation where we select, allegedly get no rows, insert, and hit a duplicate key constraint -- which would be a really annoying problem to figure out.

Reply:
I did this initially, and it's a dev-ux problem. I was actually considering changing the way we do this in entirety - not making it an abstraction layer of `database.Store`.

It's always problematic in our code to get `ErrNoRows` then hit a constraint on insert. We'd have to make everything an upsert, which sounds like a weird constraint.

Reply:
@kylecarbs what were you considering instead? Adding the `Cipher` as a field accessible in the enterprise coderd API struct, and calling it at the handler level? Or something else?

Reply:
Discussed with Kyle -- while I agree in principle that this could reduce the overall cognitive overhead, I think this might be better done as a refactor in a follow-up PR instead.